Create a default warning for qos incompatibility

mm318 opened the pull request:
Closes https://github.com/ros2/rclcpp/issues/1048

This pull request adds a default callback for incompatible offered/requested QoS events. In the case that the underlying middleware doesn't support such type of events, the default callback will not be registered and a warning will be emitted.

mm318 commented:
Hi @ivanpauno and @wjwwood, can you take a look at this pull request? It's for addressing the comment

> I think unless the user defines their own callback it would be nice to have a default callback which logs a warning when this event occurs. It would be ok, in this default case, to ignore such an error as "unimplemented" I think. Though I'd like to see a notice or warning that this is the case, something like (once) "[WARN] the rmw implementation being used does not support ON_REQUESTED_INCOMPATIBLE_QOS events, you will not be notified when publishers and subscribers on the same topic do not have matching QoS".

at https://github.com/ros2/ros2/issues/822#issuecomment-598024481.

ivanpauno commented on rclcpp/include/rclcpp/publisher.hpp:
```diff
@@ -88,6 +88,19 @@ class Publisher : public PublisherBase
+          rclcpp::get_logger("rclcpp::Publisher"),
+          "This rmw implementation does not support ON_OFFERED_INCOMPATIBLE_QOS "
+          "events, you will not be notified when Publishers offer an incompatible "
+          "QoS profile to Subscriptions on the same topic.");
```
I don't like to log a warning knowing that this will always be shown with `fastrtps` (until they support this events).

wjwwood replied:
Why?

We can add a way to suppress this warning, either programmatically or via an environment variable.

@mm318 Does fastrtps not support this already?

mm318 replied:
Correct, @wjwwood, Fast-RTPS currently still does not support this type of callback (tracked at https://github.com/eProsima/Fast-RTPS/issues/852).

mm318 replied:
> We can add a way to suppress this warning, either programmatically or via an environment variable.

Is there a standard way in ROS2 to suppress warnings?

wjwwood replied:
Not really no.

mm318 replied:
This will essentially be controlled by `options_.use_default_callbacks`.

wjwwood commented:
> There may be valid reasons why people create publishers/subscriptions in the same topic that do not match.

They can suppress the warning by defining their own callback for this case. We could also have an option to disable this when creating the pub/sub.

> If there's enough debugging tools to check if a publisher/subscription pair matched, I don't think this is needed.

I guess, but this would be nice to have still imo. As a hint to people who don't understand why their system is not working.

wjwwood commented on rclcpp/include/rclcpp/publisher.hpp:
```diff
@@ -88,6 +88,19 @@ class Publisher : public PublisherBase
+          RCL_PUBLISHER_OFFERED_INCOMPATIBLE_QOS);
+      } catch (UnsupportedEventTypeException & /*exc*/) {
+        RCLCPP_WARN_ONCE(
+          rclcpp::get_logger("rclcpp::Publisher"),
```
This should use the node's logger no?

mm318 replied:
I am unsure how to get the node's logger from a `Publisher`/`Subscription` class. The `Publisher` and `Subscription` only has access to the `NodeBaseInterface`. How do you get access to the `NodeLoggingInterface`?

ivanpauno replied:
I think that adding the logging interface to the publisher can be done in a follow-up.
For example, we're doing this currently in the code: https://github.com/ros2/rclcpp/blob/57fddca648a2c0cc422717bc2360bef2d0d90b77/rclcpp/src/rclcpp/publisher_base.cpp#L103.

wjwwood replied:
> I think that adding the logging interface to the publisher can be done in a follow-up.

I actually disagree, it will require an API change for the publisher class if we now want to pass in the logging interface, which I think we should. Better to do that now...

----

If not, you can copy what the `NodeLogger` does in the meantime:

https://github.com/ros2/rclcpp/blob/c1a7a655374a8ed68ca3f3de54adc4875c5e885d/rclcpp/src/rclcpp/node_interfaces/node_logging.cpp#L38

mm318 replied:
Done.

wjwwood commented on rclcpp/include/rclcpp/subscription.hpp:
```diff
@@ -122,6 +122,19 @@ class Subscription : public SubscriptionBase
+          RCL_SUBSCRIPTION_REQUESTED_INCOMPATIBLE_QOS);
+      } catch (UnsupportedEventTypeException & /*exc*/) {
+        RCLCPP_WARN_ONCE(
+          rclcpp::get_logger("rclcpp::Subscription"),
```
Same.

wjwwood replied:
And more cases below.

mm318 replied:
Done.

mm318 commented:
> * Adding a `options_.use_default_callbacks`, default to `true`.

I have implemented avoiding default callbacks this way.

I have addressed the suggestions made here, and applied them to `rclpy` (in https://github.com/ros2/rclpy/pull/536) as well.

mm318 commented:
@ros2/aws-oncall - please run this CI job
Gist: `https://gist.githubusercontent.com/mm318/972b53569da7d10427bf13d8992290c2/raw/0ab88f1d350b57d36fad04bb0f6358aa8b87cb14/ros2_incompatible_qos_callbacks.repos`
BUILD args: `--packages-up-to rclcpp rclpy`
TEST args: `--packages-select rclcpp rclpy`
Job: `ci_launcher`

piraka9011 commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=10133)](http://ci.ros2.org/job/ci_linux/10133/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=5700)](http://ci.ros2.org/job/ci_linux-aarch64/5700/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=8234)](http://ci.ros2.org/job/ci_osx/8234/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10028)](http://ci.ros2.org/job/ci_windows/10028/)

mm318 commented:
> Edit: ament_export_warnings only, no test failures.

Thanks for the clarification!

piraka9011 commented:
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=10029)](http://ci.ros2.org/job/ci_windows/10029/)

piraka9011 commented:
@mm318 mac OS has test failures tho.

mm318 commented:
I believe the Windows and macOS test failures are unrelated to the changes in this pull request.
